Improve voxel memory usage via `Texture3D`

jjhembd opened the pull request:
<!--
Thanks for the Pull Request!

Please review [Contribution Guide](https://github.com/CesiumGS/cesium/blob/main/CONTRIBUTING.md) before opening your first Pull Request.

To ensure your Pull Request is reviewed and accepted quickly, please refer to our [Pull Request Guidelines](https://github.com/CesiumGS/cesium/blob/main/CONTRIBUTING.md#pull-request-guidelines).

-->

# Description

This PR improves memory efficiency in `VoxelPrimitive` by reworking `Megatexture` to use the `Texture3D` class from https://github.com/CesiumGS/cesium/pull/12611.

## Megatextures

Unlike `Cesium3DTileset` which renders each tile with separate `DrawCommand`s, voxel tilesets are rendered in a single draw command for all tiles. This is necessary because of the raymarching approach: foreground tiles affect the rendering of background tiles. As a result, all tiles in the scene must be loaded into a single texture, via the `Megatexture` class. Different tiles are assigned different regions within the same `Megatexture`.

### Previous `Megatexture` implementation

Prior to this PR, the `Megatexture` implementation was backed by a 2D texture. 3D voxel tiles were split into 2D slices, which were then arranged across the 2D texture. This structure made `VoxelPrimitive` compatible with WebGL1 contexts, but had some drawbacks:

-  Shader complexity: To rendering a voxel with linear interpolation, we couldn't depend on the hardware sampling, but had to explicitly sample 4 points from two different slices, and interpolate them in the shader. 
- Memory inefficiency: the 2D texture was always sized as a square with the width and height being a power of two. This restriction, combined with the complexity of slicing 3D tiles onto a 2D surface, resulted in some datasets allocating as much as 75% more texture memory than needed.

### New `Megatexture` implementation

This PR reworks `Megatexture` to use `Texture3D`, and removes restrictions on the size of the texture. This simplifies the shader code, since we can directly use built-in graphics APIs for both linear and nearest sampling. Also, the texture can be allocated to more closely fit the size of the actual data.

## How 3D Textures are sized

The data for each tile is like a 3D box, and the `Megatexture` is like a bin into which we are packing the boxes. [Bin packing in general is a hard optimization problem](https://en.wikipedia.org/wiki/Bin_packing_problem). The approach used here starts from some simplifying assumptions:
- All tiles are the same size (it is an implicit octree)
- For a given maximum number of tiles, allocating <25% unused space is "good enough"
- For a given hard memory limit, utilizing >80% of the limit is "good enough"

All sizing is done based on the maximum number of tiles, which is either a value from the `VoxelProvider`, or the available memory divided by the memory size of one tile.

We first check for a special case: if all of the tiles can fit in a single 1x1xN stack, without the long dimension exceeding `GL_MAX_3D_TEXTURE_SIZE`, then the texture is allocated to that size. This guarantees no wasted space, because this shape can be made to fit the tiles exactly. We make sure to stack the tiles along the smallest tile dimension, to increase the chances of achieving this optimal case.

If more than one row of tiles is needed (`GL_MAX_3D_TEXTURE_SIZE` tends to be smallish), we then proceed as follows:
1. Find a factorization of the tile count into a product of small primes (2, 3, 5, or 7). The tile count is rounded up to the nearest such composite number, as needed. (There are some further simplifications--we are not checking every possible product of those primes).
2. Distribute the factors into 3 dimensions, starting from the axis of the smallest tile dimension, with the constraint `tileCount[i] * tileDimension[i] < GL_MAX_3D_TEXTURE_SIZE`
3. Compute the resulting texture size `textureSize[i] = tileCount[i] * tileDimension[i]`

## Other changes

- Updated `ContextLimits` documentation and specs to assume the default WebGL2
- Added `ContextLimits.maximum3DTextureSize`
- Added `Texture3D.prototype.copyFrom` method, following the similar method from `Texture`
- Cleaned up `Voxels` Sandcastle example
- Removed some WebGL1 workarounds from voxel shaders (Voxels require WebGL2 since https://github.com/CesiumGS/cesium/pull/12933)

## Issue number and link

Resolves https://github.com/CesiumGS/cesium/issues/12570

## Testing plan

Run all specs locally. Changes in `ShaderBuilder` should not affect non-voxel rendering, but we should verify this.

Load all voxel-related Sandcastles, and verify:
- Rendering is unchanged
- Framerate is not affected (let me know if it improves!)
- Picking works
- Nearest and linear sampling both work as expected (Open the VoxelInspector to toggle this)

# Author checklist

- [x] I have submitted a Contributor License Agreement
- [x] I have added my name to `CONTRIBUTORS.md`
- [x] I have updated `CHANGES.md` with a short summary of my change
- [x] I have added or updated unit tests to ensure consistent code coverage
- [x] I have updated the inline documentation, and included code examples where relevant
- [x] I have performed a self-review of my code

donmccurdy commented on packages/engine/Source/Renderer/ContextLimits.js:
```diff
@@ -31,11 +32,11 @@ const ContextLimits = {
   /**
    * The maximum number of texture units that can be used from the vertex and fragment
-   * shader with this WebGL implementation.  The minimum is eight.  If both shaders access the
+   * shader with this WebGL implementation.  The minimum is 32.  If both shaders access the
```
When defaulting to a WebGL2 context, is there still a fallback path for devices that only support WebGL 1? I understand voxel rendering will require WebGL 2, but other CesiumJS APIs can run with just WebGL 1?

If not, just wondering if the comments should reflect the minimums for browsers and hardware that CesiumJS supports, rather than the minimums for the preferred/default WebGL 2.

(or if deprecating WebGL 1 support more broadly is on the roadmap, maybe that changes things!)

donmccurdy reviewed (commented):
Still reading through things, but just a couple initial comments!

donmccurdy commented on packages/engine/Source/Scene/Megatexture.js:
```diff
@@ -33,52 +31,33 @@ function Megatexture(
   channelCount,
   componentType,
-  availableTextureMemoryBytes,
+  availableTextureMemoryBytes = 134217728,
```
Perhaps worth keeping the explanation of the default from below:

```suggestion
  availableTextureMemoryBytes = 134217728, // 1024x1024 @ 128bpp
```

donmccurdy commented on packages/engine/Source/Scene/Megatexture.js:
```diff
@@ -389,130 +321,173 @@ Megatexture.prototype.isFull = function () {
+    throw new DeveloperError("n and maxN must be integers");
+  } else if (n < 1) {
+    throw new DeveloperError("n and maxN must be at least 1");
+  }
```
Would it make sense to gate these checks behind a `//>>includeStart('debug', pragmas.debug);`? Asking more for my own understanding of CesiumJS best practice than an actual preference. :)

donmccurdy commented:
I'm not 100% sure I'm measuring the right thing, but opening the "Voxel Rendering" sandcastle and logging `megatexture.textureMemoryByteLength`, I'm seeing allocations of 64 MB before and 512 MB after. Other sandcastles (like "Voxel Picking" and "Voxels in 3D Tiles") have decreased, but "Voxel Rendering" is the only example I found with a large voxel set.

Is this expected and/or is there a better way to check total memory allocations? I haven't ventured into setting up something like [webgl-memory](https://greggman.github.io/webgl-memory/) in a sandcastle quite yet.

javagl commented:
> I haven't ventured into setting up something like [webgl-memory](https://greggman.github.io/webgl-memory/) in a sandcastle quite yet.

See the "Testing Plan" of https://github.com/CesiumGS/cesium/pull/12370 (I think that I did only use it in a sandcastle by "hacking" it into the local build, but think that this snippet there was supposed to work in an actual sandcastle)

javagl commented on packages/engine/Source/Scene/Megatexture.js:
```diff
@@ -389,130 +321,173 @@ Megatexture.prototype.isFull = function () {
-    } else {
-      textureDimension *= 2;
+  // Find a nearby number with no prime factor larger than 7.
+  const factors = findFactorsOfNearbyComposite(tileCount, maxTileCount);
```
I know, it's none of my business, but I hope that it's not toooo distracting: I was curious and pulled that function into a sandcastle, and tried to feed it with some of these numbers...

https://sandcastle.cesium.com/#c=lVhtc9o4EP4rWz50ICEQ4NLMXEKmubSZ6UzbZC65T8AHYURQzpYYWSaQTv77rSRblmxDc51M8cvusy96drVWJHiqICFblmTJ6Msj3apM0gf2SmEMw7NPF1M+5f2joymHI7gRyTpTFAiMvoCyorBgCeUpExzUiiiIBFeE8RTvKDyxDeXAs2ROJYglKBbTtAtCAknRKN/ZJ/ouIhyWTMELUyvGjTbZEBaTeUydrYQmQu56xpnPayJJAr9uiFQ0ZYSP3gzal8KfFB5XnnupdkBwaoSA+W5txJbGaQhr377BfIfo91Q+kGSNehqyVDQvYY03BgJSI9QM5KLJc/zDxPKXQbjmkK01TMwSzIGw8YdRQ6oXBf02RpttTHRsNyLjamY8VUKRuJp/yNNrbqjKkSRFW5ikIJ8aI7KLvmhcc+tHf8ojw6MnqhyH3DogkZYZj5S+bKN4ZZW6+lGYZfPoQL66BYoJFe868Es/sk6oun1OX+AGg8qSXhleu3OhlTTBAfp9uGV8YRJTzdcLLdjZ86zgK10m37hdwrF+VQ0EExsG29vWH+3qj14vSkNYm49FqBjKD6JWvWUshGxbi4d41a+6aTJn42ZLaAfYlzDI8whg2VBLpVF8sxnT+nlOF3TJOF203ZJ0HJCqup4wXsp1g+isX2+AtUib9H1Z3xNcuwchlSO1X/NLKRJIExLjQqKEgJjIJ7z0VjJFXbrw2sYY7ubPNFI9ypVkNG2Hq9PpaY08+23ShXkHxldAJoMZnMAcf1yWc+8esXqJlIQ/Mf4E6EtZiATLmj/FusziLMH7WKCI6X5bZlqWvi4DCMLzgzAoNwbkO+VPaoVhVCObnM60k0dlVh0TGvQvx437Qrm0FXJMmsydzmbakRr6xXsxBg6j4c1g9m6c4V6coYejkymwg8fiqT1tPdx/vfl2/R1urh++fpi2OhfvLo6inRBsPUTOd0VP0bsbcAFriVqwJJHCvdBQUurtk8O5t6b2teYjltfi1t7dLX8aRL0Xi5QpeqiaiupYx8yUR0odKONYDKZiVpL6JdOFfylds5yFdgcp+r7egnJMS8ncay2Zc0WTmpJoVWHpEuNsx1QBw3BOL/DnEkb4c3zs+GSDnmjedyE2LGlcL+YvVtgfr3X1XOvCCRpl43jTz210AjRt/dblHfczZ/gWO0n+pp0nsdtsvLOPlBp8ZmHvpVhkkbpbFpCe4Q5WaOxXSb8f8tJ3ctrq+k539qhYPCMdht2E7UD3Ce2JTKs1vyrIaKjzN+5QGwpZikNFnsmefelY8mxZ8owsKSTyjMCzRxko3k6esfeO/UTgk9z5N68u95fuWzDpXq/XUiBpiJ52uS4WqksUC5lg7Zql073ZFEFQzCm0h10YdeEMM4gN4LxTTGqSbRCtcWzjZtAqjGBdztFq6cGiUQmp9xOuvbKz2i8rhsVnNhCHoIegdCWyeIG9RwHdRrQAdWOfhZ3M7NBHt2sclblyO1AYYxgi2hQpdXlhaYN5hOH5qOiGwYNdjZvi+pkvNffHh+KV27y8UsdR8MMYB75yl1pJ8WIGwC90Q2OBU/ZXKfX01OLGfZPHJMPan7s1SItmnw8j2ggPBqT/B4vfRTEleD1wwPq/FLcDXKx26W1E0Nrgz4LdOV0np13I/1zv04LDmuCgWXDUhDjwBd/KbQdLfMjDBmou9XN0tePNpmtJN0xk6b14ofJu+fgi9FcjHB1ZEE9S4qS4yHArwbXBKamu6MnqRZAbI2uw2tajExha22YbyWW+8QXdFs5GlMVtZ6nvhHyXOZa+YxkqNoRwXEE/cveOcSHKlU9V0+L+SWleNhbdfEVZYfw2VpQsrGhg6OSkTo1AIKTJqVtUNFgPo7riJonN9BgcRsIEnGE2/2hE1MtiIAf72bkP9dMh1EGFoTUq70M9/72v2l3He7M1uKZ0eMs3huy+76SCj098ERRDYxEF2j45g9LTrc63ivNrN8TvBAWWEMZxbPO/e31DmJVKZWoiN6hdjW0K4eNHCOyPZsj00+o34v7Ulr2z2cqnRiuDw1aqtPitlbNGK8P3xDII2+P+SqpRqDbe+cZzq7qNFXvmGAZBW8UAyil04rbbc+tN0zDt67hZyR+vC1NH40AWx2nNiiA5rB50rl3EWp6BBN/Nzecso2HX/Gm6WsXKKYmNPh/ADxxnoNxp/u/C96E4IBiM3GPcBbJY2Vm7fiTVdAxVPYOC3x1CQXAKVQRXDMnWgTJkpXcp+7C3zQ9m/HOhUnBXCu7qgrtS8LUUfK0LvpaC+hvu0Zw4jKGtzBGUOXR6rTqNk70Rm7aO1fZ42jIXu+LiVV/YD0J961DtcN/qti5TtYvplc7LZ4b7nVSQybjd6/UVxZziKJv251n0L1W9KDVql/1C5XLBNsAW42krMgS6sefIVKLNKCZpim+WWRzrb7dp6+qyj/KBWizIAmv/bkNlTHZaZDW4+m4f9nq9yz7e1rWUEPGcSA/xPw

To my understanding, it's putting "three and a half" tiles along one dimension. Did I miss anything here?

jjhembd replied:
Thanks for checking, @javagl! That was clearly a bug in the `getDimensionFromFactors` helper, which would have been an issue in contexts with small maximum dimension, with small tilesets/small memory limits, where the desired tile count may include only one factor of two. I just pushed a fix. Here is an [updated Sandcastle](https://sandcastle.cesium.com/#c=tRlrbxq59q940dUKEgoBtq20CdFm00aq1G2qm6z2Q0CqGUxwdsaDPB4Cqfjve47t8diDIb3S3aoqM+Pzfvs0yUWhSEY3PCuz0Yd7tlGlZHf8hZExGb59dz4RE9E/OZkIckKu82xVKkYoGX0gyoCSOc+YKHguiFpSRZJcKMpFAW+MPPI1E0SU2YxJki+I4ikruiSXhBbAVGzNF3xLqCALrsgzV0suNDZdU57SWcocr4xludz2tDC/raikGfl+TaViBaditNPUPlTyFOR+6YlXoAC5YBqIcF+sdb5haRGSNac7MtsC9a9M3tFsBXhIskbUh2QFL5oEKTRQnJDTxtr4D63L75rClSDlCsmkPAMb5Eb/UGtSoFNAbs00zuMBdbvOS6GmWlKVK5o27U+sefULU5aSZMALjBTYE2kkxunzqM+NHP2JSHQcPTLlYsj5AQJpUYpE4WMbwBte6uKn0Mr60xF7dSsqWlV465Dv+MkIofb5C/ZMrkGpMuvV6rU754iEAU5Iv09uuJhrwzTt9cyq6Ox5XOAI0+STMC4c41FTETBsqGxvs/9pu//p5bxmBLl5X6kKqvxB1bK3SPNctg3HY3HVb4qpLWf05gvSDmhfkIG1IyEmGvZMqRF3xmKIb206Zwsu2LztXNJxhFRT9IyLGq4baGfk2hHIRRbD92F9ScB3d7lULqj9nF/IPCNFRlNwJEDkJKXyER49TxaAy+Ze2RiT29kTS1SPCSU5K9qhdzo9xLDWb9MumXXI+JLQh8GUvCEz+HFWttLdQ/ZSKal45OKRgCx1IlJIa/GYYpqlZQbvaQ4guvptuC5Z+FwrEKjnK6GpXGsin5l4VEtQo6nZw9kUhTypreoiIYJ/MY72hdq1jeB4iLE7m05RkD3q5z9KY+BoRE4G0x+mMzxIZ+jRQWPmUMHT/LE9ad19/Xj96eozub66+/jTpNU5/+HkqMoJhdJD5Wxb1RTsbkTkZCUBiyxooqAX6pCU2D4Fee/51BxjPEJ6zW/M2+3ii6aIvTgvuGLHs8lX6Jsl+Cv5z3f7uPvmhendKuU6iwrmeHMBOaMTaymZn1ld8jdjK26D1TSaqj1gp7I0TeRa5RDShhTGPqPJshHMCzBHO2WKcND67Bx+LsgIfk5PXdgZ2zxgenRJqoMp6lbu+zQso1eYZFeYX0E9jU5BfcujE1BD7jfOPdD2HOMbKDj2pG2N2I0z70Ri7hsSRv/g767r86m+3viei8Y+Qk2NWF9lPi8TdbuoRPIodKAQpH4y9vth+PtKTlqBMJ0DKIaehg7NFqPtiB4COqAZosWPqmDWofdfaIRrRsoCZhfriZ45dFH2ZKLsCaKsgrAWIU9eyJHq9OEJSvzYNwR8scLvvPQ/XCF2wUB9tVrJHIKO4lAtMNkYVgKoFxRKhHYdtgCdREHNKEh72CWjLnkLFoQ6875TDYSSr4FadDoUep6rmEBez4BrLcE8igSh+4VceWlrsJ+XHJJX9ylHAWetYpmX6RxKnCJsk7CKqJsuDdmHqZkt2WYFE7lQrtGFOoYqAs+8YM4uvIiwBzLCTqRu5jxaPIVOzi/W1cKfUqoj1yO9UgET509jmCvrZriU+bOeMz+wNUtzGOY/SolDWkto8bUdsxJqx8z5oKh6ip15kIkI5rD/jSxcv1JG4XngCOM/BXQdcFa7ljahwG3waxXdNlwfzrrE/nW1EwGHe4CDOOAoRnHgA+7q7gYpPhRhAdaP+B1E7Xgj8EqyNc/L4mv+zOTt4v45x8spOTkxRDxICQPpvIRWBL6BYWwf0YNFJ8i1htW02kaiN2RoeOs2ZGE+iTnbVMImjKdtx6nvgHyRBaS+izJAjKhw2qB+4t5dxIVULv1Q1SXuz4LZtDHU9WXNAMMVXDE6N6ABozdv9kMjAAjD5Mw5FRjuq9H0uDZiPDwGxymBAd6CNX+JUkS3aJKDw9F5iOq7Y1QHjQjdC+VDVN+/LiuK6+J+F9ukxK6bUJhpUfBHoZ+EuQjMKLYxuxiQLKNc4OwV1EvTAv6yMxYOLzHyhjYQs4Ndvb/B4TQvq8JdjXZV5Q8GtaOdBqu7EUkbrPh/lfpa7YjJsPx5mlm7HWpp9XLg3lNxj6xeaNnFVHUzA2Fqyv9eXzvuNk8Iv9UdH0SNN7qB/sHmBA6CEhstzQG2X/KCgo4N1OcKl69uSN+gQnYFocJFkpZz0NdQW5mvYIWhNhb4I8sLpfd4+gyPjC1hpjOXJyycBhtusAPy889WIrpp93o9n1uvSHnC2oNOB2rrmSt7SCA2O/uoHXs9bhrSVowrmNfqcK38/szhM8aS3cAlpZQQJZUr/UoSEbPjD5pB+bcXSHudXHAJFoJJH6NG6pVhLufGVJFNSGOn0AnuuNjADxltr6M7hf3VW5Bs/WbXRlNH0C7Hpryi84IoGk0DT71aduu5Ks7lXZTL4DiXZst4lcvbKJfhj+gyCEenw11Wtxe/ELwSvt+rEaeqO5ArwcgFCtQ33AdXst4baWIXdR/H3aP8q3vF6mQcwMJVHaMiMA7fV9piV7rWa9hgdRdf9Y6GXf0Xc8ggNha1Rnt7uT+yUQW4M/vn3Jeh2lEORu4zTIhlqsw9fH8rHtuEN9fg5LU9OAkW4ZVy1QXaCFCrrHCCNR97G7sb9lfTNeC2BtzuA25rwJca8GUf8KUGxP3QvW5kY9JWeguu994vTaHh1q/BJq1TtTmdtPTDtnp4wQezbMJXRxWptLqti0JtU3aJVvmNwyQsFSllCmW/rxhYFMaVoj8rk7+Z6iWFRrroVygXc74mfD6etBIdPtfmP7KYBI5JCu0XThZlmuJWaNK6vOgDfICW5hSHpts1kyndIshycPnZfIQaetGH130slefpjEqP4j8) demonstrating the fixed behavior.

The current code can still be broken if `GL_MAX_3D_TEXTURE_SIZE` is small and the size of each tile is large. However, it should throw an error with a reasonably clear error message in that case.

javagl replied:
Admittedly, I didn't do the math of this factor computation, e.g. what exactly is done there with these primes (and why...).  And I'm lacking some context (e.g.  was surprised that this `bytesPerSample` was only relevant for the _total_ size, and I don't know what "usual" tile dimensions are, and what usual maximum texture sizes are...). 

But I was somewhat curious: Given that the tile sizes are all the same, I think that this _might_ not even an NP-complete problem. 

First, I thought that one could compute the "wasted space" somehow directly. But the maximum texture size, the _different_ tile dimensions, and the resulting (different!) maximum numbers of tiles along each dimension make this more tricky. 

I still think that  one could map this to something that _can_ be solved deterministically, or at least _approximated_ very well with standard algorithms. Specifically: I have a veeery vague feeling that there is some "shortest path computation" hidden in that. 

But I didn't go to the level  of pen+paper right now. As a "quicker shot", I thought that ~"some greedy approach" could do it. These _can_ be pretty good for this sort of problem, and often only deliver "bad" results for "very unusual/extreme" configurations. But ... (coming back to the lack of context:) ... I don't know what "common" or "uncommon" configurations are.

I casually hacked around a bit: I created one exhaustive search (searching the configuration that minimizes the "wasted space"), and an additional simple "greedy" one (that just greedily fills the available space based on the sorted dimensions), and added some corresponding "test functions" around that. It's not really productive in some way, and certainly doesn't count as "Cesium time" - rather some recreational thing - but will dump the result here...

https://sandcastle.cesium.com/#c=7Rr9b9s29l9h88POTlQ7TtZ0WOxgubQ9BOiSoMndcIiDA23RsTqJMvTh2B78v997pEiRFOWkxda7G24YGkt833xffNQ05XlBErqKkjI5fnfHVkWZsdtow8iIHL05OR3zMe/v74852ScXabIoC0YoOX5HCglKwihhPI9SToo5Lcg05QWNeA5PjDxGS8YJL5MJy0g6I0UUszwgaUZoDkz5Wr7BpynlZBYV5Ckq5hEX2HRJo5hOYqZ5JSxJs3VPCPPTgmY0Ib9d0KxgeUT58VZQe6fkycnd3BAvRwFSzgQQiUyxlumKxblNVq5uyWQN1G9YdkuTBeAhyRpRLJIFPAgSJBdAfkJam8rGPwtd/ioonHNSLpBMHCVgg1Tqb2tNctwUkFsw9fO4R90u0pIXD0LSIi1o7NqfVOYVD6yoKGUMeIGRLHsijanc9NC751KO/phPhR89skL7kN6H6yx6jDjIMSKzkk8LdJUOoDm7FeAr29ri1Q67BYqKUHnMu+Q3fCNlKRwxgD1nT+QCdCuTXq1lpyt8nJB+n3yIeChs45rsiSkH7RkcYAkj5ZLLXRzhkqsD2NbWs7dqvlo3X21Oa0YQnndKS1DjZ1rMe7M4TbOO5LjLtfqumIjSFcSjGelYpIdkUJmQEOkPDSsKxK00GOJX5gzZLOIs7OjN6GpChSt5EvEaLrCUk3JtCUQj8+GbsKYksHW3aVZotzajfpalCckTGsM+AkRKYpo9wk9jI3PAZaGROEbkevKZTYse40UWsbxjb063hxiV8Ts0IJMuGZ0Rej94IK/JBP4oI1fC3UH40iyjHALhkYAodSRSiGv+GGOcxWUCz3EKICL9rSKRs/B3Lb+lnamDoHIhiHxk/LGYgxauYveHDyjjfm1U7Qge/OHIWxjqnXV8497H7vDhAQVpUD99KY2BpuFZGTy8mM5RK50jgw4aM4UUHqePnfHe7c37i8vzj+Ti/Pb9q/Fe9/TFsaGSCYWkQ7PJWmUULG+Ep2SRARaZ0WkBxVB4ZIb1k5O3xp7KZXRHiK7wg3y6nl0JiliM0zwq2K5gUsGxiCMRHTnTRCMOsSACZp4xM2IC8itji6jyQllCVOLHGlTRlC5ZSY2Qla+gUzM6nTteOgM9OzErSATqHJ7CnyE5hj8HB9qfpNL36PcBiYWXePcrMjfLzo7nGD3nGDhWmvT2N/2KR9eihtw/aLtDQdOMP0AiqVY6lREDP/Num1Mi8QdJ9iZLw3JaXM8USYNxFyI0NqOk37f90hRyvBeYQndbUCQ9AW2r7aOtibYBtWiGaP4l5YzCdT5BfVoyUubQVVSW7MlF7SWfpZd8Bi9REJVFyGfDZYhavf8MqXdkGgLeVMJvjbhsD92t1eqeLxZZCk5Dsd3lGCwMQxQCmULsiq3D3CyCwArmnHSOAnIckDdgQUgAb7uqVcuiJVDz9m1cdFqKCcTlBLjWEoReJHC9K3JuhJ3EfppHEHyigGgK2P7k87SMQ8g9BWGrKVNEdd8nyd4/yK6PrRbQK/NCVyBbR1tF4JnmTNslyj3sgQyvekXdBe7MalwE11W11dzsHtSSLl5GqEMT+GoErV5dpeZZ+iRav3dsyeIU2uz3WYa90x4X4gs7JiXE/kTvQa6SfdWLIBNu9UdfRhYORjGj8HugCeM/OZQD2KxOLe2UArfBj8q7K3e9PwxI9b/OfQh41AAc+AGPfRQHJuC2LjsQ4kfcTqDiJ74HUbtGZ7rI2DJKy/wmfWLZ9ezuKcVjI9nfl0QMyAwaxbAUpwAOTVIT0YDFTciWAlbQ6kiJXpMjyVuUkQrmkodspYSdsijuaE59DWSKzCH0tZcBokeFA4f6vn7WHmdTOTNdVaS4v+esChtJXRyjJDAcjgtGQwlqMXr9uukaFoDtJod6U4FhUw13x4UR/e4x2E0JDPAGrPm9lyJuiyA5aPfONqonu6gOHA9tuHIb1bfPy4riar8XpUEnpd0lXzCSdV9DWcdOWLCCwRtEFrbpnFboYaozuUL/Gtj0u1aAJTTi0LaZJ16TEVjFiUx0ZA/a2UiakHz3HbH4Hz+Apx+6R8R209a508/lxMtlsJuL6xbPcnnj5XL0El0Gdnpsj6SGCzXaO5N5xRXTmKqZIzKw0iooUHeh97rcvpXS+JppE0f3SmZ7rVjtjyxYaKfRKyzjRE2lK+yGrtV46Bcobiy8XdApa5vt1McU9YRTEg0k3oJl8bRDYUedcU6ISd6dpBgmC9fN9bW5vmmumzMWcc65q0YlABpi2g9xMBNaYEoMS3ZLkmLtBTGFKTZeEK886ACFmBmJKZEJIxrokQm6b+th1tSQiYJaj1PaYZ/EZgKwYPBa4uqMAXtkHdnqUwEh0g1Ijn7wY31gdsFqIeDpoH70wleJXOGZFqvQV/AP/JIPa/NhIx/kIVa+0saShNtEFKoLDPzVBlVZSsDJ391G5Mj3jcCZRUX+R4xBn4mjP/XQ0nVPvyehL11QjichvHHAzauXTOe0XqMbyVkwjnCMAaMP35glukToJGygONZ3UeprBnnB4KKP98Qwx329w8IKrGsekbXHCr9xTwWV0UU9ktDPD1hcJ+oGX4m5/mrMjcSUBQ06xvM4T8kcDmioCnbooqkUV1d4OSRPuwmeYWV7bfgbNmDg8/ppYD1BQzE0Mtn/kkvKUWA14hO01FDBnPuhbVzCvj34Ct+Cc5Eoc1kplypYfK3y5o4LpverOYVzdrRk3+yKCXdVqvVK5/GmoQP12sPzmQANTIoVb2FUkQmdrrXk1f3LnzFmsc2Fk9Avqie5EnOq3s317eXd5T/e/+vy6sPl1eXdP5vAFymfRVgHbfvo5nklum74MxwRI8ThjdE1a+i1hF5b0HDSIWt7HqngNxJ+Y8EfAfzGgpe+hOUV6+kGDyzNJKLOu84QNltDQB6sDkS7c7Cu/m66JlJd9a1es/U2dBWQdUA2KnhNIrotbOn52/JN0FysOxX7vSWjuWzLgyarpBnW++2Yi1huo7swL4B2lYYQFsa2ftA/t/aQ+auun5tXBT10Tkc8XxO28mOvPdiefm3tx954sDdt3d5zE/UdiftvGWPi3Pb/pL39FrfRegRaJqos7PJJzGJ6pGIeHU+/6EKPyus4zzXe/aFzk1e86NrOVr++X0JCSjV91aYnaVqTAPjUN7m1euYEWb3ta9jfObSVnF8Y0xrty4JZo311FOsIXWTQLX5ieRkXHU4TFuBkHx4CI687LIzzpypeiCmO5npI4J6sJVXSMTvQLt4vygXTnV9e3KoNaY5vAEPShZ/9lv1oTnRqrLWn/WnB2tRYG0/r4zVGU0Nx0+q+VUZpVGcnw7lbZc/hRBvebcwrspLfsbz4w7/cqvX+VHIRhABftLmJc7JRZjGcz4vmHLgRzX7lR9s1sxBX5u3r7eKrQVjgzsH+S+qVLfO7lOX8L+KYGsjLZPzEE1pR2Kler+d+JOPmTun5xieBOz8Z7NSfgDUU9qr7vLKWqvb9CUpmnSWfOW1+e+l0w7SjnfoWUuFfsxCM99Jqx+oMrfZwV2E4bRJi2rw1qdrkX0jsUVikJiQt9AyRrf4OxMx9t+ILNpEBG/NUq1fzV56T74MfgsGJuaVODrJvgHZlGoA8rP5zqqAatg7eHMuqXuVsN/87frDTBeysBMbBTzh+b9uQHwLyH7PONwgYMJtjJ3i5F+wN82IdszME+ylKFvgNa5nFnV6vXzDgSIFkf1JOf2VFb5oL7xz2FcowjJYkCkfjvakw6YX88p5lUFimMc1zWJmVcSxmcntnwz7AW2hxSkOQ6HrJspiuEWQ+OPsoX0IqH/bhsYlVpGk8oZlB8d8

(This also contains the original function, but still the one without the fix)

As the name `runSingleTest` suggests: I considered to add more test cases, with a mix of extreme tile sizes like (1,1,128) and un-extreme ones like (16,16,16), and of course, some `tileCount` values that are either prime (like 1259) or highly composite (like 1260 - yeah...). 

But ... it's ~2am here after all 🤓

javagl replied:
OK, that table formatting looks horrible (it looked OK in the preview). Just for completeness...

[Output00002.csv](https://github.com/user-attachments/files/24395203/Output00002.csv)

jjhembd replied:
Thanks again @javagl for finding the weak points. 2 quick observations from the [latest problem case](https://sandcastle.cesium.com/#c=tVl7b9rIFv8qZ1G1goRCIG1ztQnRZtNGqtRtqk2u7h+A1MEMYbL2GI3HBBLx3ffMy56xDWmlu1Uj/DjzO+/HjKOUZxISsmFJnpx+vKcbmQt6x54pjGD4/sP5REy4+usfHakfOILrNFnlkgKB048gzQKYs4TyjKUc5JJIiFIuCeMZ3lF4YGvKgefJjApIFyBZTLMupAJIhqz51jxRdxHhsGASnphcMq5XkzVhMZnFtOCV0CQV254V5/cVESSBl2siJM0Y4ac7jffRSZTB/dITMFMipJxqImC+YOt0Q+OsCmze72C2RfxvVNyRZIUrFWi5VL+EFd5oEMg00T6oQidr7z+1Rn9ojCsO+UoBxSxBS6TGCqHukCkHoeya7T4uY6XhdZpzOdXSylSSuOoHsGbWN1QWWIIiPzRWYFeFEhn3zxu972Tpq59Ix9YDlUVcFT7B4FrkPJLqsq2XVHzWNQ9Dm9uHB6zXLbG04vq+Ay/msRFI1mXh9AmuUck86ZXqtjvnZplLAYB+H24Yn2uDVe34RF309gJu+FKl02du3DsyL6uqoclDA/Q29Ufb+qPnc58Z5vG9Ux3V+pPIZW8Rp6loO66HIq9fFdZa09mBLaAdcLiAQWFbABMxNfPaxbtSzhQTI04f2t8r7H6DNy+VR7tuqNObF/92971zXjpHiWedOKcLxum8XcRBx5NTVi2UMF5Shgyd6jvAwkCbMXx6X9kiZO5SIYsc8wvRQqQJZAmJMX6QIoWYiAe8DAIow9V07lWzEdzOHmkke5RLwWjWDkOi01MrCoe3SRdmHRhdAhkPpvAWZvgTOLaQ8x4LCxGC8AfGHwClKisEwYrDH2KV/XGe4H2cIokuzxumK6q6LlUJFA3V0TjXGuYL5Q9yiQpVdRyfTJWwR6WdvRBsQLgYNbYw3+mVqBw3sTyZTpUwNfzzH0cZFCgNbwbTn0Aa7kUaBkh+Rk1ad98+XX+++gLXV3effpm0Ouc/kZ1hjSNYFYmYbV2hUy0ZeAorgWthQSKJDVwHrFA9n8NZ4GdDoKIVU3F+Y+5uF181phoh0oxJ+lreBeXCQqoyYS/DAqBSbRUznWsZLSRgHDNLp99SUD//uvA3pStmA9l0R9fRVIMtUE1cWzUVrQ02lRmURMtaqC/QNO2YSmCo/8k5/lzAKf4cH3shaew0VgnUhVgHWqO7WejrsMpfqUS8UjkYlPvGga5vuXQqeEqCm8Jd2K8L5jdYoOybtjVnt5l9pzEevyto5S/1i7Xc4+Se3oSebMwNRTc1on0T6TyP5O3CieVhdLBgxGHC9vtheviqTlqBQJ29iwympq8asAm/AN5PtkdDtbD5lR/mOiT/wr69ppBnOIhZz/Tc6yL6Hk30PWL0ORprH3gMQhHc+/EjtoiRbxZ8UqixqxWLQ1Vl17B3uFqtRIqhSdT+gavUpKp+YJUhWFi0a1Uz0QkXVJoM2sMunHbhPdoWq9NZp5x6BVsj3p4hmOuh1THCSjBDzqUU8z3LMMi/wpWX6Gb905Jhuuu+V2CoETJbpnk8x/IogW4iWsIWY7QBHk/NEE03K9yCcFm0zlDXUFXkmma0sA/LGgRAGO6N3sVwfbD4cp3MX4tA4P5E5F56ndcrMDhU/zLCwdlvsUuRPulR+iNd0zjFPcwnIdTs2eJaEW3TJMeKMys8kpU9ys5YihGvjJY/B437z5gSvB544OYnwy6GDmz7ckcEuQ5+K3PBBvX4pAv2v1eBFfGwgXiwj/i0GXkQEu/83olFYsjDgq4v1XMUvRPM/CtB1yzNs2/pExW3i/unVO3c4ejIwAS0AsfjeY5tDj2Ho2B9aUCtXCTWmlrjtY1cb2HoJNAtzlJ95nO6cUJHlMXtglu/IApF51g2imjEpQ2qHFfwj4p7Ly5DnMswqHW5/G9GbZoZDnoXa8gZSkLJ3BEH7N6+PbB7qSr+5iV4UnY0F3TB62oAnnhhggLXTVGPIu2OfWE3eA0PTfkePfNuD65yswYeHI7/fdgfDmMPahnQkC77sM9+RG4legX73Q9g/+cQ9jHU01yHxm7/MVnTWQG2IZJl7IHrK242UTOimrk97xE0IYyr2TToDK7p/c/OoGqsa2Jg0BHOjr7l8Zwa49PctSk3/Lo+VxlkX+muqpsZwbQJs/9XayuVbzCcKvCedtZ6+9t4ec5z7ylaA9anlvb00e1uURwf+9/r5YfdF4gRtvfDw7rxSjewQuUwDF8FjaOx5QTrwxIetCo1NviccQPbDTm4xZh9QeAwHsX5HDU3eCvzFO0x1GZD3yRpJvWxrX6nXhmr4qzrNp6qFZj1FyMYwK+/WqnIpt3r9Xx+vSxmEW0POh3sFideKVYQTbsMf3HHHjfUTWrryhXOrmUIuzh4YvhYRZc9cI1yITBqStf69aZB2E51DA8am92I2235ggm0Fu6NVCwJfVKcirkxW8N5U+W8plM5M1Djyj4DNs4vhfL+GWuQiv36hKJM37DwcmTKvXJnEFun04rvXm8D3oTZzOlDI6fBa5zqzexVTu8bOQ1/TKdBfXA8NBMUzckvG6+E+Es54LlahTlVGTxRofLcYFwUujMnW9MxiL+q2I2GByOO4dEooB6zqYqZwGCs2QwWIdS+PI8PDlObz/7P0Mr4NxyYtDNrK0f2ziD2FOXAyTpSnth/56Eo7hD53cmJD4ajcx5Lc+RR/3rS/MWk+rEEXv9aApXPJaWybso1gvhGkGrWNY97G/vNwP9s4ZNuS9JtnXTrkz6XpM910mefVB3U3es+OYK21N9J9JeR57r4k5ZpqJPWsdwcT1r6YusuntWFOfdTtwWuwWl1WxeZ3Mb0Utnod4YbBiEhFzF2k76kaGOci7L+LI/+prIXZWoZv+i7JRdztgY2H01akY6ua/NJlArkGcXY4fHNIo9jdTQ3aV1e9JE+WBanRM1nt2sqYrJVJMvB5RfzEMvxRR9v66tkmsYzIjzEfwA):

- The context's texture dimension limit `MAX_3D_TEXTURE_SIZE` may imply a smaller effective memory limit than `availableTextureMemoryBytes`. In the given Sandcastle, the maximum dimension limits us to 16 MB, even though we thought we had 100 MB available. We should be using the minimum of `bytesPerSample * maxDimension ** 3` and `availableTextureMemoryBytes`
- If the tile count is approximated as a power of 2, then the current approach is limited to power-of-two tile counts along each dimension. This is why we only allocate 8 tiles along Y and Z, when we really have room for 15 and 12.

javagl replied:
I think that I _roughly_ understand why that case is not working, but may not fully understand why certain limitations exist. 

And as mentioned above: I don't know which configurations _really_ appear in practice. So all this is no reason to hold up this PR. (In the hope that when an issue is opened for some error message like ~"WebGL error: negative texture size", it will be possible to 1. quickly zoom to the relevant code part, 2. find out why a certain case caused a negative texture size, and 3. fix it as a appropriate).

In the last comment, I mentioned the 'greedy2' result, but not what this does: It starts with the maximum number of tiles that _do_ fit (along each dimension), and then always decreases the tile count by 1 (along one dimension) that causes the greatest reduction in "wasted space". It is guaranteed to find **a** solution (if one exists), often finds the _ideal_ solution, often one that is better than the current one, and I haven't seen a case where it finds one that is _worse_, but I didn't _really_ (really really) systematically, exhaustively, and fairly compare things here. (For cases like a maximum size of 2048, and a single tile of size 1x1x1, would mean that it decrements the size 2047+2047+2047 times, but maybe that's not an issue either).

The core of the approach (other approaches and sanity checks omitted here): 

https://sandcastle.cesium.com/#c=vVdrb+JGFP0rUz5BwhpI2q1UQ1Rt0lZI3bBaUFdVQNVgX4OV8RjNjJPYFf997/g9ftCXVCQSe+bcuee+ByfkUpGAvvlBFNw+bOBNRQLWfgJkQW6+e29vxZbrrxdxR/khJ04YnCIFX6hU4K5P1IGh3idE+Qwe/AC4RJgcV4v3YcRV7V1+AlEC0/UR+TPbdlI67hsqN4+z3mwDEbcRsYlI2ojEQKhQUbZBRG4tar3SB+OfBrCgYxBvMFJxJ8gkpZJOUA8vqcGalNKklImKJLh6vwJfmRYZaN8FynLdaTguol/T2CI8VfIuk84RAjBBeA5J186tFDmJ0I0cNaRlXBkocsIDZ/kpXijIMFPmkdAjFZIg7mpBvBx4NrSeehTmOfkBpFpHjgNShmLYyEfiREIAJmLlBINe4PMvhd2PUbAHYX1arZeb5W8//bF8/Hn5uNz8bnfAS4XaX9wFz+e5Y3IzNdrHzamN/+YFDYsBP6gjrl1f14zPfMKxDFHiybKsHL6zC4Tee/J35F3lTUJ8jwwLr2vAiNwtOuysNJQh7i3n7NNZ1NVWvbSzD4dXcg/SjwLrngqFT5Tfppyeprtxxn5WPNzsRoawR5mEamFkV8/awpz0vHK+YRkxYljL0NZ2PWaaSA11Lh7PXQnYPsHMyMy5B1BlKy2d94sAcOMb1Fim7aXGuY9V2iXWNDgxyBfpC/UZ3TPID/8IQSjiDxra7LdGW51MyGb1sCKrwFfaO9uB5yu5HRDnCM4zOYIATLV6C8CJULSgp8IjH6k6Wh4LsbY6B8ak1bOr6P5j2fg/yCaF7K5WsHkhoUGFbfnm6xFfyFCJCPoK8d/2F6M+v8kKs5awewH02W4kHKkxreXm2RgRjdxKoV11VzBoClh6pOV6sC71NOgety3BuCY46xCM+wSTmuBNh2BiDpimfFZndses8bn6DDJi2PloAGM8QL/UwjBuqDJuGyEDi4WHVJZcY2X8sB2M7PY+Fkt+MqZKxo1IzL7RdlCoHJljvDnm/zpI7SsGymRn4+OkN0btW0clF3eUVq9cUsklHWXV45a2pdonrdVRz/AWEd9gYf1vzdCk/zni3OcHzDep+iNvsirtq+VUj6DJPBU0l/oEL1iXnnJh/5IR2eWv4G90qSxts/BXw+rSLCtvCj23hM6w/Z3ANUJX3QTq1b4dHDIaVQXmvC7Vfn8SrjENNCdMxcbvkUb0++r4+zGZ4fd2Znq0kQLVtS3bvhRoxE7zT6u3ZJFckG+n0+LAopKak6kRh4sBaOZF6aqGh+ziulNlmBtyKEpoMB7MpYoZ3GliP/rBKRT4o0WwoWVNFCAPitom+8h5BmU5UgeGzyeFyNz1X/Bnx2I7cFI334dcUbxYC31nYVRK3PEixvT83w7u5hPEG2IspC6yXb2AYDTWkOPs7tdsEa868wm+tqVUGLI9FbUTvwI

jjhembd commented on packages/engine/Source/Scene/Megatexture.js:
```diff
@@ -389,130 +321,173 @@ Megatexture.prototype.isFull = function () {
+ * @returns {Cartesian3} The computed 3D texture dimensions.
  */
-Megatexture.getApproximateTextureMemoryByteLength = function (
+Megatexture.get3DTextureDimension = function (
```
I added a comment and a `@private` tag. I know the tag is technically redundant, but it's to communicate the intent to the reader. I don't see much use of `@ignore` elsewhere in the repo.

jjhembd commented on packages/engine/Source/Renderer/Texture3D.js:
```diff
@@ -68,7 +68,7 @@ function Texture3D(options) {
   } = options;
 
-  if (!context.webgl2) {
+  if (!context.webgl2 && !defined(context.options.getWebGLStub)) {
```
The `Context.webgl2` flag is set [as follows](https://github.com/CesiumGS/cesium/blob/95214adcd8d14a72756422e3f5d846711e7ee22e/packages/engine/Source/Renderer/Context.js#L65):

```javascript
const webgl2 = webgl2Supported && glContext instanceof WebGL2RenderingContext;
```

It will be hard to make the stub pass that `instanceof` check. This makes me think a replacement of the `.webgl2` flag with a `.webgl1` flag is the best route. However, that would be a breaking change... so maybe it's better to just accept that support for both WebGL1 and WebGL2 will be messy, but temporary (we will drop WebGL1 _someday_).

I added a comment clarifying the check in `Texture3D`.

jjhembd commented:
Thanks a lot for the feedback @donmccurdy! I think I addressed the code comments.

For your testing results:

- I couldn't reproduce the issue with the cylinder in "Voxel Picking". It looks like an LOD didn't load immediately. Did you run both versions with the same viewport size? The fact that it resolved when you moved the camera makes me think it might be working as intended.
- The change in the translucency in "Voxel Rendering" was intentional--see https://github.com/CesiumGS/cesium/pull/13084/commits/51aa022990defaf10d1cd9b873a9fe0648277a27. This is restoring the example to the intended behavior, as it was before the shader's ray distance was changed from arbitrary "shape" coordinates to actual distances in meters (See https://github.com/CesiumGS/cesium/pull/12933 -- I neglected to update the Sandcastle in that PR.
- I still need to investigate the memory usage for "Voxel Rendering" -- thanks for flagging that!

donmccurdy reviewed (commented):
Minor comments left only, code changes look great!

@jjhembd I'm still seeing a rendering difference with that cylinder in the "Voxel Picking" demo. The difference goes away as soon as the mouse "hovers" over the region, no camera movement involved. Not sure if that's a blocker, but it's unexpected to me. Let me know if it'd be helpful to do a quick video call this week.

jjhembd commented on packages/engine/Source/Scene/VoxelTraversal.js:
```diff
@@ -96,12 +84,14 @@ function VoxelTraversal(
       this.megatextures[i].textureMemoryByteLength;
   }
 
+  // TODO: this number actually varies per megatexture, depending on bytes per sample!
```
Sorry I missed this. It was an existing issue which I noticed while editing the file. I removed the comments and documented the issue in https://github.com/CesiumGS/cesium/issues/13169.

ggetz commented:
@jjhembd, are you on track to work with @donmccurdy to get this merged ahead of Monday's release?
